src/types/block.rs

```diff
@@ -1,5 +1,104 @@
-use serde::{Serialize, Serializer};
-use types::H256;
+use serde::{Serialize, Serializer, Deserialize, Deserializer};
+use types::{Bytes, Transaction, U64, U256, H256, H160, H2048};
+
+/// Block transactions: either hashes or full transactions,
+/// based on flag in request.
+#[derive(Debug, Clone, PartialEq)]
+pub enum BlockTransactions {
+  /// Full transactions.
+  Full(Vec<Transaction>),
+  /// Transaction hashes.
+  Hashes(Vec<H256>),
+  /// No transactions.
+  None,
+}
+
+impl BlockTransactions {
+  /// Convert to `Option<Vec<Transaction>>`. Some on `Full` variant, None otherwise.
+  pub fn full(self) -> Option<Vec<Transaction>> {
+    match self {
+      BlockTransactions::Full(txs) => Some(txs),
+      _ => None,
+    }
+  }
+
+  /// Convert to `Option<Vec<Transaction>>`. Some on `Hashes` variant, None otherwise.
+  pub fn hashes(self) -> Option<Vec<H256>> {
+    match self {
+      BlockTransactions::Hashes(txs) => Some(txs),
+      _ => None,
+    }
+  }
+}
+
+impl Deserialize for BlockTransactions {
+  fn deserialize<D>(deserializer: &mut D) -> Result<Self, D::Error>
+    where D: Deserializer
+  {
+    let res = Vec::<Transaction>::deserialize(deserializer).map(BlockTransactions::Full)
+      .or_else(|_| Vec::<H256>::deserialize(deserializer).map(BlockTransactions::Hashes));
```

Comment:
so this line seems not to work as expected -- it only works if the type of transaction you're trying to get comes first. My guess is that the calls of the first attempt's `visit_*` leave the deserializer in an invalid state to deserialize the other type. We might just want to make `Block` generic over a transaction type and change the `eth` API to have two calls: 

`block(&self, BlockId) -> Block<H256>` and 

`block_with_txs(&self, BlockId) -> Block<Transaction>`

(uncle functions will return `Block<H256>`)